compiler/rustc_typeck/src/check/upvar.rs

```diff
@@ -602,7 +602,47 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
             }
         }
 
-        debug!("For closure={:?}, min_captures={:#?}", closure_def_id, root_var_min_capture_list);
+        debug!(
+            "For closure={:?}, min_captures before sorting={:?}",
+            closure_def_id, root_var_min_capture_list
+        );
+
+        // Now that we have the minimized list of captures, sort the captures by field id.
+        // This causes the closure to capture the upvars in the same order as the fields are
+        // declared which is also the drop order. Thus, in situations where we capture all the
+        // fields of some type, the obserable drop order will remain the same as it previously
+        // was even though we're dropping each capture individually.
+        // See https://github.com/rust-lang/project-rfc-2229/issues/42 and
+        // `src/test/ui/closures/2229_closure_analysis/preserve_field_drop_order.rs`.
+        for (_, captures) in &mut root_var_min_capture_list {
+            captures.sort_by(|capture1, capture2| {
+                for (p1, p2) in capture1.place.projections.iter().zip(&capture2.place.projections) {
+                    match (p1.kind, p2.kind) {
+                        // Paths are the same, continue to next loop.
+                        (ProjectionKind::Deref, ProjectionKind::Deref) => {}
+                        (ProjectionKind::Field(i1, _), ProjectionKind::Field(i2, _))
+                            if i1 == i2 => {}
+
+                        // Fields are different, compare them.
+                        (ProjectionKind::Field(i1, _), ProjectionKind::Field(i2, _)) => {
+                            return i1.cmp(&i2);
+                        }
+
+                        (l, r) => bug!("ProjectionKinds were different: ({:?}, {:?})", l, r),
```

Comment:
a) Might be a good idea to mention why this is a bug
b) I'd prefer if we matched on `(Projection::Index | Subslice, _) | (_,  Projection::Index | Subslice) => bug!("We don't capture index/subslice)`

Reply:
I implemented both but in order to make the `match` exhaustive, it's pretty ugly 🙁

Reply:
Little bit of duplication but this works

```
(l @ ProjectionKind::Slice | ProjectionKind::Index, r)  => bug!(Didn't expect index/slice , l, r)
(l, r  @ ProjectionKind::Slice | ProjectionKind::Index)  => bug!(Didn't expect index/slice , l, r)
(l, r) => bug!("Projections must be of same kind, found", l, r);
```

Reply:
I don't think this really gets us very much benefit since we have the wildcard pattern match at the end, if new variants were to be added to `ProjectionKind`, this code would continue to compile. At the same time, we have more code than just falling back to the wildcard match which would show the `ProjectionKind` anyway. I think Niko basically raised the same point [here](https://github.com/rust-lang/rust/pull/89208#discussion_r715630997).